### ISSUE TITLE
lebedev_mapping_ access is thread-unsafe (and also never initialized?)

### ISSUE THREAD

**Issue by tallakahath** [psi4/psi4#2735]
I've been battling intermittent (read: ~1/1000 runs) hangs of DFT runs that I think I've traced back to some rather mystifying bits of code in cubature.cc... (h/t for @rmcgibbo for lots of help here)

The problems begin here:
https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libfock/cubature.cc#L5106

(Don't mind the mismatch between line numbers I quote and line numbers in my gdb outputs -- there's some edits in a local version of psi4 we're using but I pinkie-promise they're not at all related to the code in-question! ...but to be very sure, I'll check that the same things happen with a truly vanilla build when I get home and have access to my "normal"/vanilla psi4 build.)

Which is where my runs always hang (on the [] access, specifically) when they do. If I pop open gdb, attach, and check what lebedev_mapping_ looks like, I see something somewhat disturbing:

```
#4  psi::SphericalGrid::build (scheme="", npoints=1852534357, points=0x1b2) at /tmp/garden-install.1c2be761ae20/psi4-1.6.1-desres02-c7/psi4/src/psi4/libfock/cubature.cc:5142
5142    /tmp/garden-install.1c2be761ae20/psi4-1.6.1-desres02-c7/psi4/src/psi4/libfock/cubature.cc: No such file or directory.
(gdb) print lebedev_mapping_
$5 = std::map with 6 elements = {
  [26] = 0,
  [26] = 0,
  [26] = 0,
  [26] = 0,
  [26] = 0,
  [26] = 0
}
```

Which is obviously not correct in two ways:
1. There shouldn't be multiple entries for 26
2. The entry for 26 should be 3

The threads are all deadlocking on the `[]` access, probably because multiple threads all tried to access an _unset_ entry of 26, wrote 0 to it in the process, and ran into each other. From what I can find `[]` isn't thread-safe (because of situations like this), and `.at()` should be used instead. Since `lebedev_mapping_` is a static member of `SphericalGrid`, all threads within their nice shiny `new SphericalGrid()`s inside `SphericalGrid::build` are still accessing the same global map, and mayhem is happening in the process.

...but the more disturbing problem, IMO, is that `lebedev_mapping_` appears uninitialized up until this point. I suspect in an ideal codebase all entries in `lebedev_mapping_` that would ever be accessed would already exist, so the difference between `[]` and `.at()` is just the slight performance hit in the latter of checking that the entry exists.

There _is_ a function that's supposed to take care of this... https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libfock/cubature.cc#L5125

But from what I can tell, it never gets called. In the attached minimal example I set a breakpoint at `void SphericalGrid::initialize_lebedev()` and it's _never hit_. I also set a breakpoint at `s->order_ = lebedev_mapping_[npoints];` (the line where things hang on an Unlucky Run), and that line is definitely hit (so my breakpoints are firing correctly).

I'm not sure what the consequences of this all are, as I seem to be getting sensible energies (though I suppose I haven't compared to a known good literature benchmark calculated without psi4).

I don't think I'm the first person to run into concerns in this part of the code, though, given this comment:
https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libfock/cubature.cc#L3812

Here's a minimal file to just check that `initialize_lebedev` is never called:
```
set freeze_core 1
set {
basis STO-3G
guess sad
fail_on_maxiter false
dft_nuclear_scheme SBECKE
dft_radial_scheme EM
dft_radial_points 99
dft_spherical_points 590
dft_vv10_radial_points 50
dft_vv10_spherical_points 194
dft_pruning_scheme robust
dft_vv10_postscf False
wcombine False
}
molecule mol {
0 1
C 0.9314972462818318 9.553873448534286 1.4618215727211896
C 0.407816021141855 8.469506309378442 2.18786320317928
C 0.8158282265296265 7.159078273321417 1.778073368593668
C 1.6705486753474763 6.941146237307343 0.7209312466052733
C 2.2362488662446047 8.008509869086277 0.003971742160492786
C 1.911850027633602 9.341079539288122 0.4146901215657727
H 0.614495362639488 10.566900793575911 1.4754165926364617
H -0.36704844148631577 8.614110059906363 3.04421928826377
H 0.4567894628980789 6.157197048912012 2.369426098541459
H 1.8364925212621395 5.872552432013023 0.3910028702339573
H 2.9471948988553818 7.873466587177986 -0.782748863606265
H 2.4659886000145983 10.188822366489614 -0.03294075189628659
units angstrom
no_reorient
no_com
symmetry c1
}
energy('pbe')
```

And below is the full file that is an example of one I ran into the hang with (note that if I run it again, it usually runs just fine... I have only noticed this horrible hang behavior because I am running a _very large_ number of calcs on a daily basis, and eventually get unlucky!). Note that when things hang, they hang on the very first step of the first SCF run and _only_ there. I think that is because if you aren't unlucky and `lebedev_mapping_` gets successfully populated, there's no more hangs that can happen.

(Sorry the below is LONG, I wanted to reproduce exactly all settings that went into a run, in case they are important. It is hard to figure out the "minimal example" as the crashes are rare _enough_ to make it very hard to reproduce them "in captivity").

```
np.set_printoptions(threshold=sys.maxsize)
set DF_INTS_NUM_THREADS = 16
set freeze_core 1
set {
basis def2-tzvppd
guess sad
fail_on_maxiter false
dft_nuclear_scheme SBECKE
dft_radial_scheme EM
dft_radial_points 99
dft_spherical_points 590
dft_vv10_radial_points 50
dft_vv10_spherical_points 194
dft_pruning_scheme robust
dft_vv10_postscf False
}
set {
wcombine False
}
molecule mol {
0 1
@C 0.265781511352271 0.841876264095443 -0.6914658815200431
@C 1.4413385866197737 0.3288357318169406 -1.1836757404873801
@C 2.2289711018973604 -0.6237867174726249 -0.5257029978620363
@C 1.8767103570771173 -0.9962343566137659 0.6986750981164255
@C 0.6545489119459532 -0.6147226134119103 1.2098027282027453
@C -0.21957387152459434 0.2564142719883635 0.5001192004375371
@H -0.3723596175396408 1.4255337233361631 -1.1634648379610923
@H 1.9177612974037643 0.7339585876082476 -2.2134537293183647
@H 3.30111923095952 -0.8660596018835396 -0.9695609764149236
@H 2.5278636235464713 -1.4596675993689474 1.3422341203604324
@H 0.5072122755030857 -0.940559021032073 2.215163414817783
@H -0.998224655511569 0.6769295855057103 1.0319031933299063
@C 1.317102898709397 -4.2559640682725135 -5.346326015162741
@C 2.6128913245771566 -4.000991995558086 -5.963528222785021
@C 2.6860602549649695 -4.080680715147077 -7.357260511037285
@C 1.61182676831893 -4.323741601719452 -8.13762758477727
@C 0.36704188574136254 -4.4498345447883 -7.491026236218776
@C 0.18935870670297217 -4.482238426772551 -6.1101157069039
@H 1.4643016900247157 -4.417830669371555 -4.268504690440918
@H 3.5167101959887903 -4.155258555420228 -5.441815579461341
@H 3.7606934564339944 -3.859729234284167 -7.848947020120773
@H 1.6613567438629124 -4.4495016057397585 -9.191513591480888
@H -0.43042396762295976 -4.715135428367288 -8.131538403210945
@H -0.7857432342736794 -4.7178574202550685 -5.387218656011778
@C -1.626234859047421 4.512413614610493 -9.823046494766203
@C -0.9826706183621148 3.6939244652828753 -8.821257560942346
@C -0.6986319218032849 2.3111056533569103 -9.261042373944903
@C -1.177145703486778 1.829413070447374 -10.546452404078376
@C -1.7690317253158327 2.758224993404836 -11.422010911826131
@C -1.898144381103452 4.0463911069083895 -11.094379169774317
@H -1.7012274142018224 5.627199302988766 -9.77335737394089
@H -0.7354458407999775 3.808094895046909 -7.760202384000629
@H -0.3592431605443381 1.650666094497526 -8.464737898970688
@H -0.8600381404496702 0.8579427307272174 -10.923006258691348
@H -2.0234174460096264 2.2620312875347652 -12.390674641213696
@H -2.3635026438425073 4.6343554516716985 -11.896791070451545
C 0.9314972462818318 9.553873448534286 1.4618215727211896
C 0.407816021141855 8.469506309378442 2.18786320317928
C 0.8158282265296265 7.159078273321417 1.778073368593668
C 1.6705486753474763 6.941146237307343 0.7209312466052733
C 2.2362488662446047 8.008509869086277 0.003971742160492786
C 1.911850027633602 9.341079539288122 0.4146901215657727
H 0.614495362639488 10.566900793575911 1.4754165926364617
H -0.36704844148631577 8.614110059906363 3.04421928826377
H 0.4567894628980789 6.157197048912012 2.369426098541459
H 1.8364925212621395 5.872552432013023 0.3910028702339573
H 2.9471948988553818 7.873466587177986 -0.782748863606265
H 2.4659886000145983 10.188822366489614 -0.03294075189628659
@C -2.883378976118354 3.679902163454515 6.605611804401257
@C -1.4418573642560812 3.600920790635927 6.670521485926086
@C -0.6393020528317843 4.660543897500103 7.199305306584455
@C -1.3266187948685466 5.747101703223244 7.774753886547491
@C -2.724567555535229 5.633863788466681 7.845926916061886
@C -3.515096369644318 4.64011668323281 7.307695334883003
@H -3.522185022496425 2.9755923057599087 6.166222160695254
@H -0.7769743608626567 2.795784357233139 6.267882984917893
@H 0.4237337787110078 4.948400919867275 7.2125183897413745
@H -0.9232262610709561 6.536591753828505 8.324128824246781
@H -3.2319912639558943 6.425323678439093 8.347870777526017
@H -4.670840490698138 4.668965154614611 7.387704241614836
C 2.2081658074790274 -5.3020164768913265 6.6510953531518195
C 3.4850970832998747 -4.830430366645136 6.603095343785539
C 3.6960958670600284 -3.5170046939420203 7.135768650558369
C 2.606915315437915 -2.6915518986780596 7.483946158304935
C 1.3596547084775734 -3.2600808377424353 7.521569539699344
C 1.1711131766388743 -4.555253723319432 7.114249143031859
H 2.116450827322973 -6.447024106625211 6.2912440007746735
H 4.147926021324595 -5.527468935787702 6.298140312113919
H 4.682959113326085 -3.1084423755103856 7.144261556531443
H 2.853922074403954 -1.7713832406320649 7.868036685742114
H 0.3668345387515738 -2.7610843910048732 7.8237465009707545
H 0.1081678166328347 -4.80913158701848 7.0546213578338275
@C -1.7405771438695425 -7.80886257185175 0.5404281767388233
@C -2.6188602604225517 -6.691098033004703 0.5011602693432181
@C -3.714140810909062 -6.576587282189308 1.3810925093268906
@C -3.942255953512009 -7.605544961402358 2.270588825183478
@C -3.03532531969506 -8.58700653050214 2.445310345258381
@C -1.9513792884695083 -8.764665860879806 1.6060394926445083
@H -0.9796992846786388 -7.840283014272942 -0.17744029414230778
@H -2.495309384578162 -5.759810624845935 -0.07447588606303623
@H -4.340889311265438 -5.63947864160487 1.3227958668441364
@H -4.551193864181943 -7.442030783029586 3.213032536935569
@H -3.1972325817701313 -9.21254035719057 3.319783691154892
@H -1.0777792942495945 -9.375803514705414 1.8259033760084107
units angstrom
no_reorient
no_com
symmetry c1
}
set basis def2-tzvppd
data = dict()
mol.set_full_geometry(Matrix.from_list([[0.5022542654449073, 1.590915569963503, -1.306681140595813],
[2.7237351815665702, 0.6214094731249062, -2.2368229698847557],
[4.212144922291332, -1.1787860562025736, -0.9934346888539755],
[3.546468590124386, -1.8826100899427796, 1.3203045855360824],
[1.2369181787496408, -1.1616573819627147, 2.286195821128258],
[-0.4149344813052484, 0.4845527485033083, 0.9450883184937658],
[-0.7036576970203194, 2.6938683185233923, -2.198629899378565],
[3.624043624500105, 1.3869807173967377, -4.182821337943252],
[6.238211251248736, -1.6366154551623187, -1.8322047065489515],
[4.77696992890693, -2.7583719957972828, 2.5364548826091995],
[0.9584922877490646, -1.7773989537967476, 4.186052175296396],
[-1.8863712097657719, 1.279211522262235, 1.9500144225202807],
[2.488963756511477, -8.042606485283144, -10.103091941619423],
[4.937648996861485, -7.5607790984247085, -11.269435077539946],
[5.07591823612626, -7.711368953670002, -13.903207393384701],
[3.045911152463411, -8.17068746090052, -15.377887439428918],
[0.6936086403173598, -8.408968589548671, -14.155987978839677],
[0.35783609498195545, -8.470203051873602, -11.546445275822556],
[2.767129157973372, -8.348490030084381, -8.06630482661384],
[6.64561913009783, -7.852300646752344, -10.28354106590362],
[7.106680671332505, -7.2938311680086105, -14.832360234725817],
[3.13950924120095, -8.408339425930736, -17.36944335867427],
[-0.8133834162821838, -8.910314600134187, -15.366380553945444],
[-1.484839517054919, -8.915458419315543, -10.180367833336788],
[-3.0731384979190186, 8.52722589264645, -18.562867584573294],
[-1.8569783394210484, 6.980505564439329, -16.669760864965355],
[-1.3202229941291759, 4.367356729918863, -17.500833715309945],
[-2.224482988369945, 3.457089671956247, -19.929906630109166],
[-3.3429854666211285, 5.212289827632862, -21.584472415838455],
[-3.5869730252828074, 7.6465709851781725, -20.96533815362649],
[-3.2148538885467297, 10.633865531333795, -18.468968754840013],
[-1.3897912186070827, 7.1962564082241745, -14.664657177429172],
[-0.6788711855737655, 3.1193068418059653, -15.996036345794762],
[-1.6252365421823092, 1.621276791687995, -20.641490286499835],
[-3.823704808747945, 4.274619618775313, -23.414981571239565],
[-4.466372691691263, 8.75766256782587, -22.48167688504687],
[1.7602746813157428, 18.05420424706374, 2.7624424155127625],
[0.7706605891926988, 16.005047335515233, 4.134462252155306],
[1.541691912880146, 13.528697241336232, 3.360071696132801],
[3.1568794742631003, 13.116865379487553, 1.3623626107690086],
[4.225897903707189, 15.133890318935643, 0.007505504920960816],
[3.6128729435857183, 17.65208203760002, 0.7836507563470771],
[1.161227940241227, 19.968548485949185, 2.7881332798114595],
[-0.6936210288798643, 16.278308820606714, 5.752740718121715],
[0.8632069814922737, 11.635416117797726, 4.477566398783909],
[3.4704678951089125, 11.097515749009613, 0.7388883386847691],
[5.5693911947308194, 14.878695501158344, -1.479180976577893],
[4.660043080476915, 19.254083805126786, -0.06224899942322865],
[-5.448796578367559, 6.954007254346888, 12.48279719591343],
[-2.7247155292187717, 6.804754090772794, 12.605458716832036],
[-1.2081057907631072, 8.807151558071153, 13.604715317009585],
[-2.5069461940826483, 10.860448229462223, 14.692155531944536],
[-5.148686488002696, 10.646459583648602, 14.82665306518665],
[-6.642569440294071, 8.76854971761692, 13.809542785134276],
[-6.6559650527771, 5.623054516429932, 11.652471107312186],
[-1.468268747886573, 5.28326673868353, 11.844582222696696],
[0.8007407915165865, 9.351122493395383, 13.62968442543807],
[-1.7446447844863573, 12.352368203223108, 15.730323703929539],
[-6.107578326059717, 12.142102014322907, 15.775189493290956],
[-8.826609299232123, 8.823065427642629, 13.96073770638817],
[4.172828613898477, -10.019359049577828, 12.568748646228057],
[6.585879005168848, -9.128190457034952, 12.478041774546314],
[6.984608919104673, -6.646175650573697, 13.484648438245728],
[4.926355976270059, -5.086295938718023, 14.142608570641505],
[2.5693750230805064, -6.160659927473747, 14.21370645735928],
[2.213083164787955, -8.608181965255541, 13.443982462683802],
[3.999512419878047, -12.183109880381776, 11.88872814465905],
[7.838444165512118, -10.445402451015424, 11.901760284360375],
[8.849510177006636, -5.874104763895764, 13.500697704537037],
[5.393131101647324, -3.34742918654667, 14.868434474560834],
[0.6932168112938935, -5.217693305978344, 14.784738155333894],
[0.20440754893479765, -9.087941596751813, 13.331302279241509],
[-3.2892141006967823, -14.756611605640634, 1.0212612440672273],
[-4.948928650864744, -12.644342755401091, 0.9470556536023668],
[-7.018708920913597, -12.427948798034729, 2.6098865953999857],
[-7.4497840653131195, -14.37239700557793, 4.290791021232555],
[-5.735933553365823, -16.22709057302199, 4.620966842252425],
[-3.687572420474477, -16.562818050920185, 3.034974786431045],
[-1.8513633325344034, -14.81598763653126, -0.335313559402],
[-4.715451333061206, -10.884464610668445, -0.14073902754797188],
[-8.203091935607064, -10.657070118309775, 2.4997219071223564],
[-8.600509943381129, -14.063399990960185, 6.07175152431976],
[-6.0418939362757165, -17.409178187152243, 6.273481969281969],
[-2.0367076889241837, -17.717700841096935, 3.450457310685484]]))
mol.update_geometry()
e, wf = energy('wb97m-v',return_wfn=True,)
```

**Comment by tallakahath** [comment]
In the slightly modified codebase I work off of -- I flipped the `[]` to a `.at()` and tests now fail due to `std::out_of_range`, as expected, given that the map is starting and staying empty until accessed.

**Comment by tallakahath** [comment]
Building from ToT in conda at home, I can confirm that `initialize_lebedev` in fact is never called... it's optimized straight out of the code! (I couldn't find anywhere where it got called, so I am not surprised)

```
(base) liz@Gaston:~/Development/psi4/objdir/stage/tests$ nm ../lib/psi4/core.cpython-38-x86_64-linux-gnu.so | grep lebedev
00000000020dbe80 b _ZN3psi13SphericalGrid16lebedev_mapping_E
```

Doing the same breakpoint tests using the line-numbers in ToT cubature.cc, I can also see that `initialize_lebedev` is never called, and `lebedev_mapping_` ends up with no entries in it when it's first accessed.

Will work on a patch now.

**Comment by tallakahath** [comment]
I have a patch written -- I can see that `initialize_lebedev` actually gets called, and when I inspect `lebedev_mapping_`'s values at a breakpoint, things look reasonable.

...I also see no changes in the energy anywhere.

In fact, I don't see any evidence of any use of `order_` at all, except in one printing function. Is this intentional? 

Anyway, PR incoming. I am OK not understanding the mysteries of `order_`, as long as I no longer get 1/1000 hangs when trying to run large workflows.

**Comment by tallakahath** [comment]
Oh also h/t to @averyparr for first finding the issue (for us anyway) and identifying the somewhat strange conditions it happened under.

Avery, it's fixed!

**Comment by averyparr** [comment]
This is great! Thanks for finding/fixing the hangs! It was a huge pain to trawl through all the `stdout` files.

**Comment by JonathonMisiewicz** [comment]
Closed by #2743
